### PR TITLE
Feature/md 6332 composite action for git diff

### PR DESCRIPTION
**Pull request by felix11h**
As discussed in MD-6332.

### Usage
The new composite action will be used in the repositories listed below as indicated:

* https://github.com/moneymeets/moneymeets-app/compare/feature/MD-6332-use-composite-action-to-check-for-diff
* https://github.com/moneymeets/moneymeets-dashboard/compare/feature/MD-6332-use-composite-action-to-check-for-diff
* https://github.com/moneymeets/django-insurance-api-v2/compare/feature/MD-6332-use-composite-action-for-git-diff
* https://github.com/moneymeets/django-securities-api/compare/feature/MD-6332-use-composite-action-for-git-diff
* https://github.com/moneymeets/django-profile-api/compare/feature/MD-6332-use-composite-action-to-check-for-diff
* https://github.com/moneymeets/fondsnet-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff
* https://github.com/moneymeets/morgen-und-morgen-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff
* https://github.com/moneymeets/moneymeets-reporting/compare/feature/MD-6332-use-composite-action-to-check-for-diff

### CODE REVIEW

File: check-for-diff/action.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+name: Check for diff
+description: 'Check for changes between committed code and current state'
+
+inputs:
+  directory_to_check:
```
Comment (zyv) [inline check-for-diff/action.yml]: Screenshot 2022-10-10 at 13 39 36

Sounds a bit weird, maybe `path_to_check`?

**Comment by zyv** [comment]
I don't quite remember why did we decide to restrict this in APIs @mm-matthias ?

Screenshot 2022-10-10 at 13 45 55

Is there a good reason for that, like there were some temporary files left over or so? If yes, it would be good to leave a comment I think that explains it, and if no, then maybe the restriction can be removed?

**Review by marns93 — APPROVED** [review]
Approved from my side, but I would prefer to rename it as Yury suggested it.

File: check-for-diff/action.yml (hunk 1 above)
Comment (felix11h) [reply check-for-diff/action.yml]: Sure! 5bf923eafa2625d9fae06aab1d81f96cad23e531

**Comment by felix11h** [comment]
As dicussed in MD-5838, I have added an additional commit to provide a composite action also for Python linting (flak8e + add-trailing-comma).

File: lint-python/action.yml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+name: Lint Python
+description: 'Linting for Python projects'
+inputs:
+  exclude_trailing_comma:
```
Comment (zyv) [inline lint-python/action.yml]:
```suggestion
  exclude_trailing_comma_path:
```

Also here no strong opinion on that, but without `path` sounds like a mask or boolean.

Comment (felix11h) [reply lint-python/action.yml]: https://github.com/moneymeets/moneymeets-composite-actions/pull/4/commits/a173c6c42f5822ad1b428aff6af7a9d6bae64ce9 :)

File: lint-python/action.yml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    default: ''
+
+runs:
+  using: "composite"
+  steps:
+    - run: find ${{ inputs.exclude_trailing_comma_path }} -name '*.py' | xargs poetry run add-trailing-comma --exit-zero-even-if-changed --py36-plus
+      if: ${{ inputs.exclude_trailing_comma_path }} != ''
+      shell: bash
+
+    - run: |
```
Comment (marns93) [inline lint-python/action.yml]: I would specify a name for the step, because you execute two commands. E.g. `Run linter`. In the GitHub actions logfile you need to expand the command to what is really executed.

Otherwise you can put it to another `run` step.

**Review by marns93 — CHANGES_REQUESTED** [review]
Thank you for introducing the lint action as well. 

But now I think there are much more repositories to update. This is used in all our Python repositories (see https://github.com/search?l=YAML&q=org%3Amoneymeets+add-trailing-comma&type=Code).

File: lint-python/action.yml (hunk 3 above)
Comment (felix11h) [reply lint-python/action.yml]: Thanks for the feedback. Sorry, I don't quite understand what you are suggesting. The logfile [here](https://github.com/moneymeets/django-insurance-api-v2/actions/runs/3224685636/jobs/5276087184), for example, looks good to me. Can you give an example where you would like to add the name? Thank you!

Comment (felix11h) [reply lint-python/action.yml]: I have split up the run steps as we have discussed, @marns93. 

As for leaving out names, after looking at 
Screenshot 2022-10-11 at 10 51 52
I might still have preference to simply call the step "Run linter", but this is something we can decide in the pull requests that change the ci.yml workflows in each repository.

**Review by marns93 — APPROVED** [review]

File: check-for-diff/action.yml [hunk 4]
```diff
@@ -0,0 +1,14 @@
+name: Check for diff
+description: 'Check for changes between committed code and current state'
```
Comment (pchorus) [inline check-for-diff/action.yml]: This file has mixed single quotes and double quotes. In most cases quotes are not needed at all, like here in this line.
I would prefer removing them everywhere it is possible and use quotes only where needed and use the same type of quotes (single or double).

File: check-for-diff/action.yml [hunk 5]
```diff
@@ -0,0 +1,14 @@
+name: Check for diff
```
Comment (pchorus) [inline check-for-diff/action.yml]: Maybe `Check git diff` would be a bit more precise?

File: check-for-diff/action.yml [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+inputs:
+  path_to_check:
+    description: 'Path to check'
+    required: false
+    default: ''
+
+runs:
+  using: "composite"
+  steps:
+    - run: git diff --color-words --exit-code && test -z "$(git status --porcelain ${{ inputs.path_to_check }})"
```
Comment (pchorus) [inline check-for-diff/action.yml]:
```suggestion
    - run: git diff ${{ inputs.path_to_check }} --color-words --exit-code && test -z "$(git status --porcelain ${{ inputs.path_to_check }})"
```
The first part with `git diff` must be narrowed to the path_to_check as well, right?
Probably the default value for `path_to_check` should be `'.'` then, right?

**Review by pchorus — CHANGES_REQUESTED** [review]

File: check-for-diff/action.yml (hunk 5 above)
Comment (felix11h) [reply check-for-diff/action.yml]: Agreed! f0f5c6ba72d552a35e448b474dcca2558477b3b4

File: check-for-diff/action.yml (hunk 4 above)
Comment (felix11h) [reply check-for-diff/action.yml]: I don't have a strong opinion on this, but I tried to stay consistent with other actions in this repository. The general rule seems to be single quotes for yaml strings, double quotes within run statements. I'm happy to change it, but I would then suggest refactoring all actions in this repository. 

@marns93, what do you think? Personally, I'm ok with the way it is. I like using quotes even if they are not strictly necessary. Between single and double quotes in yaml, I believe we should use single quotes. For run statements, double quotes should be used. 

So, my suggestion would be to only change the `using: "composite"` to `using: 'composite'` in all actions, as this is indeed inconsistent with our general usage, as @pchorus points out.

File: check-for-diff/action.yml (hunk 6 above)
Comment (felix11h) [reply check-for-diff/action.yml]: I'm not sure! It wasn't part of the original requirements, and I have already implemented the use of this action in all of our CI workflows and have not encountered any problems. It doesn't seem to be needed, at least at the moment. @zyv even raised the point whether in our APIs, we really need to restrict the git status check to a subfolder at all.

**Comment by pchorus** [comment]
In all repositories where you run this action you give it a name. Those are different. Sometimes it is
`- name: Check that all code is checked in` and sometimes `- name: Check for diff`.

What about not giving them names? I guess it just shows the name of the action in the logs. So that would be sufficient and it is the same name everywhere. What do you think, @felix11h?

File: check-for-diff/action.yml (hunk 6 above)
Comment (pchorus) [reply check-for-diff/action.yml]: So from my point of view it does not make sense to narrow it down to a directory for the `git status --porcelain`, but not for `git diff`.

We want to find changes for a specific path or for the whole repository. `git status` shows you the changes, but it does not show you if there are additional files. That is why we have `git status --porcelain` additionally, but both commands should run on the same files.

Or am I missing something here?

The default value for `path_to_check` can still be empty string. That should work even after my suggested change. So I am okay with leaving that as it is.

File: check-for-diff/action.yml (hunk 4 above)
Comment (marns93) [reply check-for-diff/action.yml]: I agree, we should keep it consistent. In the past I followed to docs (https://docs.github.com/en/actions/creating-actions/creating-a-composite-action#creating-an-action-metadata-file) and there you can see mixed quotes, I think this is the reason why it is mixed for us as well. :(

So @felix11h please update the other actions in the repository as well. I thought using single quotes in the description is a good idea. Otherwise if you use some magic characters, which will be interpreted by YAML it will break your file.

Comment (marns93) [reply check-for-diff/action.yml]: I talked to @pchorus and he explained me how it is in most of our other repos. So as mentioned before, I don't have a strong opinion. I'm also fine if we decide on not using quotes, just if they are really necessary.

Comment (felix11h) [reply check-for-diff/action.yml]: I have rebased and have add a commit to remove the quotes as requested, see c7ae8fe86d84b7a41da744609c2406129706f363.

**Comment by felix11h** [comment]
> In all repositories where you run this action you give it a name. Those are different. Sometimes it is `- name: Check that all code is checked in` and sometimes `- name: Check for diff`.
> 
> What about not giving them names? I guess it just shows the name of the action in the logs. So that would be sufficient and it is the same name everywhere. What do you think, @felix11h?

I have briefly discussed this with @marns93 also. The resulting name will be

`RUN moneymeets/moneymeets-composite-actions/check-git-diff@master`

I'm ok with this, if it is preferred by others. Personally, however, I would much rather give a name `Check for diff` or similar, as it is much easier to read. As for the difference in names, I have reused the name if it was already present, but we can certainly agree a standard name such as `Check git diff`.

File: check-for-diff/action.yml (hunk 6 above)
Comment (felix11h) [reply check-for-diff/action.yml]: I can't comment on how much sense the command makes in each specific context, but I can answer the following:

- `git diff --color-words --exit-code && test -z "$(git status --porcelain path/to/dir/)"`, as it is currently implemented in the composite action here, is used exactly like this for example in the profile API, [ci.yml#L44](https://github.com/moneymeets/django-profile-api/blob/78af2ade8e0d36ea114ec59c8c34309a55a2b392/.github/workflows/ci.yml#L44), and securities API [ci.yml#L46](https://github.com/moneymeets/django-securities-api/blob/ac44ba04a45d0070345681a4a7f388760bbdbe64/.github/workflows/ci.yml#L46)
- It is imaginable that one wants to A) check that committed code has not changed anywhere, and simultaneously B) check that there are no additional files in the path to check, but allow additional files elsewhere.

Does this help you at all, @pchorus? 

Since @zyv has also asked about this in his comment above, maybe @mm-matthias can clarify why, if at all, narrowing down the path to check is really needed in our APIs.

Comment (pchorus) [reply check-for-diff/action.yml]: After discussion with @mm-matthias we agreed that both commands should run on the same files, so both of them should be narrowed to the given path, if it is provided. Therefore, please adopt the code as described above.

Furthermore, the `git status` command would be sufficient on its own and `git diff` does not find more changes, but we would like to keep it so that in the logs one can see immediately what has changed.

Comment (zyv) [reply check-for-diff/action.yml]: So did you also discuss why the restriction on the path is needed or not?

I'm file with making the action itself more flexible and allow it to be restricted if wanted, but actually I would like to avoid restrictions if possible.

Comment (mm-matthias) [reply check-for-diff/action.yml]: The other thing I suggested is to actually define paths you want to exclude from check and otherwise always check everything.

The main question is what this action should do in the first place:
1. Make sure the files in a certain path didn't change (where path can also be the whole repository). One use case is to make sure the api is always regenerated from the spec.
2. Ensure the repository in general has no unexpected changes (except maybe some anticipated changes like .tmp folders being created by some tool or something like that)

In general 1) is a subset of 2). I think Yury is aiming to achieve 2) while the people in this PR seems to aim for 1).

I'd also prefer 2) if this is easy to implement. Maybe we can try to check the whole repository and even don't implement any exceptions first. My guess is that almost all of our repositories don't have any changes in whatever path there is.

Comment (felix11h) [reply check-for-diff/action.yml]: I have checked whether repositories, that are using a git diff check, necessarily need a restriction to a specific directory:

- [x] [moneymeets-app](https://github.com/moneymeets/moneymeets-app/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before
- [x] [moneymeets-dashboard](https://github.com/moneymeets/moneymeets-dashboard/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions 
- [x] [django-insurance-api-v2](https://github.com/moneymeets/django-insurance-api-v2/compare/feature/MD-6332-use-composite-action-for-git-diff) Had restrictions, but also works without restrictions
- [x] [django-securities-api](https://github.com/moneymeets/django-securities-api/compare/feature/MD-6332-use-composite-action-for-git-diff) Had restrictions, but also works without restrictions
- [x] [django-profile-api](https://github.com/moneymeets/django-profile-api/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions
- [x] [fondsnet-sdk](https://github.com/moneymeets/fondsnet-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before
- [x] [morgen-und-morgen-sdk](https://github.com/moneymeets/morgen-und-morgen-sdk/compare/feature/MD-6332-use-composite-action-to-check-for-diff) Had restrictions, but also works without restrictions
- [x] [moneymeets-reporiting](https://github.com/moneymeets/moneymeets-reporting/compare/feature/MD-6332-use-composite-action-to-check-for-diff) No restrictions used before

In summary, currently, we do not need to restrict our diff checks to any subfolders. While this could become relevant in the future, I believe we can for now simply remove the option to restrict our checks. 

I have added a commit removing the parameter. As Pascal noted, `git diff` functionality is already covered by `git status`, but we want to keep it, as this allows us to easily see what has changed in the logs. https://github.com/moneymeets/moneymeets-composite-actions/pull/4/commits/6e9008acfc739b4c06c511a05e4a4fedee99932b

**Review by zyv — APPROVED** [review]
👍

File: create-deployment/action.yml [hunk 7]
```diff
@@ -1,4 +1,4 @@
-name: Create deployment
+name: 'Create deployment'
```
Comment (marns93) [inline create-deployment/action.yml]: As discussed, please remove quotes if not necessary.

File: create-deployment/action.yml [hunk 8]
```diff
@@ -13,7 +13,7 @@ inputs:
     required: true
 
 runs:
-  using: "composite"
+  using: 'composite'
```
Comment (marns93) [inline create-deployment/action.yml]: same here

**Review by marns93 — CHANGES_REQUESTED** [review]
Thanks for testing it in every repository. :) I think it is much simpler now.
Two last places, where you can remove the quotes.

But it is approved from my side.

**Review by mm-matthias — APPROVED** [review]
Thank you very much!

**Review by marns93 — APPROVED** [review]
:+1:
